frame/name-service/src/lib.rs

```diff
@@ -0,0 +1,797 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! ## Name Service: Register Friendly Account Aliases and Metadata
+//!
+//! # Index
+//!
+//! * [Key Terms](#key-terms)
+//! * [Goals](#goals)
+//! * [Limitations](#limitations)
+//! * [Usage](#usage)
+//!
+//! The name service pallet provides a means to register names and subnames, also termed nodes,
+//! via a commit reveal scheme. Names can act as aliases to addresses for a particular para ID, and
+//! can be used by UIs to transfer tokens between accounts in a more user-friendly manner.
+//!
+//! ## Key Terms
+//!
+//! * commitment: A hash that represents a commitment to purchase a name registration. Any account
+//!   //! can register a commitment by providing an owner address and a commitment hash - a
+//!   bake2_256 hash of the desired name and a secret.
+//! * node: Either a to-level name hash or a subnode record that exists in the service registry.
+//! * name hash: A blake2_256 hash representation of a registered name.
+//! * subnode: A child name of a registered name hash. Subnodes of a name can be registered
+//!   recursively, so the depth a subnode can be registered is unbounded.
```

Comment:
The pallet leaves the market to decide how valuable subnode depth will be. Since all subnodes require a deposit, the user is always trading subnode registrations with token value, so this should prevent any flooding of storage.